automation/scripts/clean_pipelines/clean_pipelines.py

```diff
@@ -0,0 +1,422 @@
+        formatted_date = formatted_date.rstrip("Z") + "Z"
+
+        return formatted_date
+    except (ValueError, OverflowError) as e:
+        raise ValueError(
+            f"Invalid date format: {date_input}."
+            f" Date format must adhere to the RFC 3339 standard (e.g., 'YYYY-MM-DDTHH:MM:SSZ' for UTC)."
+        ) from e
+
+
+def _get_kfp_client(
+    kfp_url=mlrun.mlconf.kfp_url, namespace: str = mlrun.mlconf.namespace
+) -> Client:
+    kfp_client = mlrun_pipelines.utils.get_client(kfp_url, namespace)
+    return kfp_client
+
+
+def _get_list_runs_filter(project_name: str, end_date: str, start_date: str) -> str:
+    filters = {
+        "predicates": [
+            {
+                "key": "created_at",
+                "op": 7,  # Operation 7 corresponds to '<=' (less than or equal)
+                "timestamp_value": end_date,
+            },
+        ]
+    }
+    if project_name != "*":
+        filters["predicates"].append(
+            {
+                "key": "name",
+                "op": 9,  # Operation 9 corresponds to substring matching
+                "string_value": project_name,
+            }
+        )
+    if start_date:
+        filters["predicates"].append(
+            {
+                "key": "created_at",
+                "op": 5,  # Operation 5 corresponds to '>=' (greater than or equal)
+                "timestamp_value": start_date,
+            }
+        )
+    return json.dumps(filters)
+
+
+def _query_and_filter_runs(
+    kfp_client: Client, project_name: str, query_filter: str
+) -> tuple[list[tuple[str, str]], set]:
+    """
+    Query the pipeline runs and filter them based on the project name.
+
+    :param kfp_client: KFP client for interacting with the pipeline API.
+    :param project_name: Name of the project for filtering the runs.
+    :param query_filter: Filter for querying the runs.
+    """
+    runs = _list_pipelines_runs(kfp_client, query_filter)
+
+    # Filter out non-project-related runs if project was provided
+    project_runs = _filter_project_runs(project_name, runs)
+
+    if project_name == "*":
+        project_names = [
+            mlrun_pipelines.mixins.PipelineProviderMixin().resolve_project_from_workflow_manifest(
+                run.workflow_manifest()
+            )
+            for run in project_runs
+        ]
+        project_names = set(project_names)
+    else:
+        project_names = {project_name}
+
+    mlrun.utils.logger.info(
+        f"Found {len(project_runs)} runs for projects", project_names=project_names
+    )
+    runs = [(run.id, run.name) for run in project_runs]
+
+    # Collect experiment IDs
+    experiment_ids = set(run.experiment_id for run in project_runs if run.experiment_id)
+
+    return runs, experiment_ids
+
+
+def _list_pipelines_runs(
+    kfp_client: Client,
+    query_filter: str,
+    page_token: str = "",
+    sort_by: str = "",
+    batch_size: int = 1000,
+) -> list[PipelineRun]:
+    runs = []
```

Comment:
I fear this can explode in memory. Maybe we should add some default limit to number of runs (10k?). If someone passes a long time range and we need to pull 500k runs this can kill the pod

Reply:
This will be paginated in a follow-up PR, where the deletion will happen after each page, so it shouldn't be a concern.